### ISSUE TITLE
when should `services.getEmitOutput(filePath)` return emitSkipped?

### ISSUE THREAD

**Issue by laurelnaiad** [TypeStrong/atom-typescript#755]
I'm having trouble understanding when `getEmitOutput(filePath)` is expected to skip emits.

It seems to just stop emitting for no particular reason and then just stay stopped.

I've traced it as far as "hasEmitBlockingDiagnostics" in ntypescript. (I don't know the tsc process enough to know if this code is from tsc or ntsc.)

Is there a reason why various outputs are being quashed by that code?

**Comment by laurelnaiad** [comment]
I think I might be on to something. The issue I was trying to solve is "why do I get emit errors when I have declaration=true in my tsconfig.json?"

It seems that setting causes problems unless you're redirecting output from sc via `outDir` parameter. I think "hasEmitBlockingDiagnostics" is the culprit: https://github.com/TypeStrong/ntypescript/blob/faf5a7eb600758e1963f9d4209f50f55d9ccc9e7/src/compiler/program.ts#L569

Is it preventing the output of d.ts files? Seemingly because it doesn't want files in the current directory to be modified or something? Which in turn is bombing the compile?

Anyway, that's my theory.

Maybe this is all well known and obvious and I'm just now catching on... or maybe I have a configuration problem or have broken something somewhere?

**Comment by basarat** [comment]
> Maybe this is all well known and obvious

no. Can you share your tsconfig.json :rose:

**Comment by laurelnaiad** [comment]
On my current codebase, I'm not longer getting this problem, I think. I have several local changes in place, a couple of which may be helping out with the particular issue I raised here. Let me try this out for a few days, and if it seems like the improvement I hope it is, I'll package them up as a single patch. I'll look to post a branch to my fork some time today if it's holding up, so you can see what I'm trying to do... :)

**Comment by laurelnaiad** [comment]
I would like to revive this issue, because it turns out it's still an issue. 

I've pushed https://github.com/laurelnaiad/at-dts-prob, a repo that demonstrates the d.ts problem, having just tested it against atom-typescript master.

@basarat, would you please take a look and let me know what you think? The readme in that repo explains how to reproduce the problem, as well as where the problem seems to live. You asked for a tsconfig.json, I'm giving you a repository with a tsconfig file in it, instead.

Thanks!

P.S. To keep the discussion here rather than on that random repository, here is what I said in its readme:

--

**at-dts-prob**

To see the problem, modify the name of the class in demo.ts and save the file.

For example, it currently exports the class Demo. Change the class name to Demoize and save the file.

Result: TS Emit Failed.

The problem seems to stem from https://github.com/TypeStrong/ntypescript/blob/faf5a7eb600758e1963f9d4209f50f55d9ccc9e7/src/compiler/program.ts#L569 and https://github.com/TypeStrong/ntypescript/blob/faf5a7eb600758e1963f9d4209f50f55d9ccc9e7/src/compiler/program.ts#L1322, in which "createEmitBlockingDiagnostics" is putting a lock on d.ts files.

In other words, the d.ts file can only be created once, after which changes to the interface it would generate result in TS Emit Failed.

Workaround: delete the d.ts file and reload Atom window.

**Comment by laurelnaiad** [comment]
Hmm... maybe I should have the project ignore its own d.ts files....

```
    "filesGlob": [
        "./**/*.ts",
        "!./**/*.d.ts"
    ],
```

seems to fix it. This might be developer error on my part?

**Comment by basarat** [comment]
> seems to fix it. This might be developer error on my part?

Definitely not. If this is the *correct* fix than most likely a developer error on *my part* that needs to be fixed. Sadly I am going to mostly away till 8th January and will not be able to provide much support :rose:

**Comment by laurelnaiad** [comment]
No worries @basarat. Hopefully you'll be doing fun/relaxing things till then!  I don't know that my "have the project ignore its own d.ts files" maneuver is correct, *per se,* but it keeps the emits from beginning to fail. :)

**Comment by basarat** [comment]
FWIW, if one calls the language service endpoint without passing it a file the language service emits *all project files* so we can have *much faster builds*. But how it interacts with babel might be a problem. Note to self : explore in future :rose:

**Comment by laurelnaiad** [comment]
@basarat in a related problem, the locks that are placed on files during transpilation -- are those locks an ntypescript thing or a typescript thing?

I ask because I am starting to become convinced that there is contention between atom-typescript and my build process for these files, but since I have emit turned off I would expect atom-typescript not to be locking any files....

So I guess that if the locking is an ntypescript thing, my second question would be, "is there any way to disable that?" :)

**Comment by basarat** [comment]
> are those locks an ntypescript thing or a typescript thing

@laurelnaiad absolutely nothing special about nTypeScript here. Also in our usage of ntypescript (which is exactly equivalent to what it would be with typescript ... except that it would be more difficult) we do not get it to do any file reading / writing for us. So if there are any locks they would be becuase of: 

* either atom internals
* Node native : fs.readFile / fs.writeFile (although I doubt it ... but you might be able to find docs confirming or denying your suspicions)

:rose:
